org.scala-ide.sdt.core/src/org/scalaide/ui/internal/editor/outline/ModelBuilder.scala

```diff
@@ -0,0 +1,417 @@
+  }
+}
+trait HasModifiers {
+  import scala.reflect.internal.Flags._
+  private[outline] var flags = 0L
+  def setFlags(f: Long) = {
+    flags = f
+  }
+  def isPrivate = (flags & PRIVATE) != 0
+  def isProtected = (flags & PROTECTED) != 0
+  def isTrait = (flags & TRAIT) != 0
+  def isParam = (flags & PARAM) != 0
+  def isSynthetic = (flags & SYNTHETIC) != 0
+  def isImplicit = (flags & IMPLICIT) != 0
+}
+sealed abstract class NodeKey
+
+sealed abstract class Node(val isLeaf: Boolean, val parent: ContainerNode) extends HasModifiers {
+  def name: String
+  var start: Int = 0
+  var end: Int = 0
+  def displayName = name
+  def key: NodeKey
+  def cp(that: Node): Boolean = {
+    start = that.start
+    end = that.end
+    val f = flags
+    flags = that.flags
+    f != that.flags
+  }
+}
+
+sealed abstract class ContainerNode(parent: ContainerNode) extends Node(false, parent) {
+  protected var _children: Map[NodeKey, Node] = new scala.collection.immutable.ListMap[NodeKey, Node]()
+  def children = _children
+  def children_=(ch: Map[NodeKey, Node]) = _children = ch
+  def addChild(ch: Node) = {
+    last = Some(ch)
+    _children += (ch.key -> ch)
+  }
+  var last: Option[Node] = None
+}
+
+case class ClassKey(name: String) extends NodeKey
+case class ClassNode(name: String, override val parent: ContainerNode) extends ContainerNode(parent) {
+  override def key = ClassKey(name)
+}
+
+case class ObjectKey(name: String) extends NodeKey
+case class ObjectNode(name: String, override val parent: ContainerNode) extends ContainerNode(parent) {
+  override def key = ObjectKey(name)
+}
+
+case class TraitKey(name: String) extends NodeKey
+case class TraitNode(name: String, override val parent: ContainerNode) extends ContainerNode(parent) {
+  override def key = ObjectKey(name)
+}
+
+case class PackageKey(name: String) extends NodeKey
+case class PackageNode(name: String, override val parent: ContainerNode) extends Node(true, parent) {
+  override def key = PackageKey(name)
+}
+
+case class RootNode() extends ContainerNode(null) {
+  override def name = "ROOT"
+  override def key = null
+  def diff(rn: RootNode): (Iterable[Node], Iterable[Node]) = {
+    val toUpdate = new MutableList[Node]
+    val toRefresh = new MutableList[Node]
+    def visitContainer(cn: ContainerNode, cn1: ContainerNode): Unit = {
+      var children = collection.immutable.ListMap[NodeKey, Node]()
+      var needUpdate = false
+
+      cn1.children.foreach(p =>
+        cn.children.get(p._1) match {
+          case Some(n) =>
+            if (visitNode(n, p._2))
+              needUpdate = true
+            children += (p._1 -> n)
+          case None =>
+            needUpdate = true
+            children += (p._1 -> p._2)
+        })
+      if (children.size != cn.children.size)
+        needUpdate = true
+      if (needUpdate)
+        toUpdate += cn
+      cn.children = children
+    }
+
+    def visitNode(n: Node, n1: Node): Boolean = {
+      val b = n.cp(n1)
```

Comment:
This has side effects. Can you please make them clean by documenting it in Scaladoc. It may also be useful to rename the `diff` method to make the side effects clear.